src/library/scala/collection/immutable/LazyList.scala

```diff
@@ -624,11 +624,47 @@ final class LazyList[+A] private(private[this] var lazyState: () => LazyList.Sta
     if (knownIsEmpty) LazyList.empty
     else super.intersect(that)
 
-  // TODO: override to detect cycles
-  override def distinct: LazyList[A] = super.distinct
-
-  // TODO: override to detect cycles
-  override def distinctBy[B](f: A => B): LazyList[A] = super.distinctBy(f)
+  /**Overrode distinctBy to detect circular lazy list so that they terminate
+  @param f The treatment to apply to each element - to check if they are distinct
+  @return A lazy list containing all the distinct elements as defined by f
+  Done as part of Bug Fix #11270
+    */
+  override def distinctBy[B](f : A => B) : LazyList[A] = if(isCyclic){
+    var set = Set[B]()
+    def isDistinct(elem : A) : Boolean = {
+      if(set.contains(f(elem))) false
+      else {
+        set = set + f(elem)
+        true
+      }
+    }
+    @tailrec
+    def internalDistinct(distinctLL : LazyList[A], slowIterator : LazyList[A], fastIterator : LazyList[A]) : LazyList[A] =
+      if(slowIterator.isEmpty) distinctLL
+      else if(fastIterator.isEmpty || fastIterator.tail.isEmpty)
+        internalDistinct(if(isDistinct(slowIterator.head)) slowIterator.head #:: distinctLL else distinctLL, slowIterator.tail, fastIterator)
+      else if(slowIterator ne fastIterator)
+        internalDistinct(if(isDistinct(slowIterator.head)) slowIterator.head #:: distinctLL else distinctLL, slowIterator.tail, fastIterator.tail.tail)
+      else
+        slowIterator.head #:: distinctLL
+
+    internalDistinct(distinctLL = LazyList(),slowIterator = this, fastIterator = if(this.isEmpty) this else this.tail).reverse //reversed to maintain original sequence
```

Comment:
This does all the slow/fast work twice, first by checking to see if it's a cyclic case, and then actually handling it.  Can we change this to just do the work on the first pass (using the same algorithm as the superclass so it's just as efficient)?